File: backend/app/elements/gui/widgets/spacer/spacer_element.py
"""The `spacer` widget: deliberate empty space between sections."""

from __future__ import annotations

from app.elements.base import StaticWidget
from app.models.graph import GuiWidgetKind


class SpacerElement(StaticWidget):
    """Nothing, on purpose.

    The grid gap is uniform, so it cannot say "this section ends here". A
    document does that with vertical space, and this is the block that holds it
    -- LaTeX's `\vspace`, and one of the twenty widgets Node-RED's dashboard
    converged on after a decade.
    """

    widget_kind = GuiWidgetKind.SPACER

File: frontend/src/elements/gui/widgets/spacer/spacerElement.ts
import type { GuiWidgetElementDefinition } from '../../../types';
import SpacerEditor from './SpacerEditor';
import SpacerWidget from '../../../../components/gui/widgets/SpacerWidget';

/** Nothing, on purpose: the block that says "this section ends here". */
export const spacerElement: GuiWidgetElementDefinition = {
  widgetKind: 'spacer',
  ports: () => ({ inputs: [], outputs: [] }),
  ConfigEditor: SpacerEditor,
  RuntimeWidget: SpacerWidget,
};
